l1-contracts/contracts/state-transition/chain-deps/facets/Mailbox.sol

```diff
@@ -213,14 +224,27 @@ contract MailboxFacet is ZkSyncHyperchainBase, IMailbox {
             ++ptr;
             settlementLayerBatchNumber = uint256(settlementLayerPackedBatchInfo >> 128);
             settlementLayerBatchRootMask = uint256(settlementLayerPackedBatchInfo & ((1 << 128) - 1));
+
+            uint256 settlementLayerChainId = uint256(_proof[ptr]);
+            ++ptr;
+
+            // Assuming that `settlementLayerChainId` is an honest chain, the `chainIdLeaf` should belong 
+            // to a chain's message root only if the chain has indeed executed its batch on top of it.
+            // 
+            // We trust all chains from the same STM.
+            // Note that the logic below will stop working in case the STM of the `settlementLayerChainId` is 
+            // changed, so it is the responsibility of the STM to ensure that gateways never leave.
+            require(IBridgehub(s.bridgehub).stateTransitionManager(settlementLayerChainId) == s.stateTransitionManager, "Mailbox: wrong STM");
```

Comment:
I would instead check that the settlementLayer is whitelisted. Also in the future we might not have this, i.e. the Gateway will have chains of other STMs settling on it.

Reply:
@kelemeno the problem with this is that any stm can whitelist a settlement layer

Reply:
And it is okay, since the assumption is that within the same STM the chains can trust each other, including trusting the gateway. 

However, we do plan to allow permissionless addition of new STMs in the future